File: scala3/src/main/scala/surd/Factoring.scala
package surd

/** Polynomial factoring over Q.
  *
  * Implements:
  *   - Rational root test
  *   - Kronecker's method for small degree
  *   - Square-free factorization (via Poly.squareFree)
  */
object Factoring:

  /** Find all rational roots of a polynomial over Q.
    * By the rational root theorem, if p/q is a root (in lowest terms),
    * then p | a_0 and q | a_n.
    */
  def rationalRoots(f: Poly[Rational])(using Field[Rational]): Vector[Rational] =
    if f.isZero then Vector.empty
    else if f.degree == 0 then Vector.empty
    else
      val a0 = f.coeffs.head
      val an = f.coeffs.last
      if a0.isZero then
        Rational.zero +: rationalRoots(Poly.divMod(f, Poly(Vector(Rational.zero, Rational.one)))._1)
      else
        val numDivisors = intDivisors(a0.num.abs)
        val denDivisors = intDivisors(an.num.abs * a0.den) // account for denominators
        val candidates = for
          p <- numDivisors
          q <- denDivisors
          if q != BigInt(0)
          sign <- Vector(BigInt(1), BigInt(-1))
        yield Rational(sign * p * an.den, q)

        candidates.distinct.filter(r => Poly.eval(f, r).isZero)

  /** Integer divisors of |n|, including 1 and |n|. */
  private def intDivisors(n: BigInt): Vector[BigInt] =
    val abs = n.abs
    if abs == BigInt(0) then Vector(BigInt(1))
    else
      (BigInt(1) to abs.min(BigInt(1000))).filter(d => abs % d == 0).toVector

  /** Factor a polynomial over Q into irreducible factors.
    * Returns (factor, multiplicity) pairs.
    *
    * Strategy: square-free decomposition, then rational root extraction.
    */
  def factor(f: Poly[Rational])(using Field[Rational]): Vector[(Poly[Rational], Int)] =
    if f.isZero then Vector.empty
    else if f.degree == 0 then Vector((f, 1))
    else
      val sff = Poly.squareFree(f)
      sff.flatMap { case (sf, mult) =>
        factorSquareFree(sf).map((fac, _) => (fac, mult))
      }

  /** Factor a square-free polynomial over Q. */
  private def factorSquareFree(f: Poly[Rational])(using Field[Rational]): Vector[(Poly[Rational], Int)] =
    if f.degree <= 1 then Vector((f, 1))
    else
      rationalRoots(f).headOption match
        case Some(r) =>
          // (x - r) is a factor
          val linearFactor = Poly(Vector(-r, Rational.one))
          val quotient = Poly.divMod(f, linearFactor)._1
          (linearFactor, 1) +: factorSquareFree(quotient)
        case None =>
          // No rational roots; for degrees 2-4 this means irreducible over Q
          // (assuming square-free). For higher degrees, we'd need Berlekamp/LLL.
          Vector((f, 1))

  /** Check if a polynomial is irreducible over Q.
    * Reliable for degrees <= 4 (no rational roots + square-free => irreducible for deg 2,3).
    * For higher degrees this is a heuristic.
    */
  def isIrreducible(f: Poly[Rational])(using Field[Rational]): Boolean =
    if f.degree <= 1 then f.degree == 1
    else
      rationalRoots(f).isEmpty && (f.degree <= 3 || !hasNontrivialFactor(f))

  /** Check for nontrivial factors by trying small-degree divisions. */
  private def hasNontrivialFactor(f: Poly[Rational])(using Field[Rational]): Boolean =
    // Try to find degree-2 factors by Kronecker's method (simplified)
    val n = f.degree
    if n <= 3 then false
    else
      // Evaluate at 0, 1, -1, 2, -2 and check if any divisor combination works
      val pts = Vector(Rational.zero, Rational.one, -Rational.one, Rational(2))
      val vals = pts.map(p => Poly.eval(f, p))
      // If f(0) = 0, it has a rational root (already checked)
      // For a proper Kronecker implementation we'd interpolate candidate factors.
      // For now, this is conservative (returns false = assume irreducible).
      false

File: haskell/algebraic-polynomials/src/Math/Polynomial/Factoring.hs
-- |
-- Module      : Math.Polynomial.Factoring
-- Description : Polynomial factoring over the rationals
-- Stability   : experimental
--
-- Factoring of univariate polynomials over \(\mathbb{Q}\) using the
-- rational root theorem and Kronecker's method (for quadratic factors).
-- Sufficient for the radical denesting and trig evaluation use cases,
-- where the polynomials to be factored are typically cyclotomic or
-- low-degree minimal polynomials.
module Math.Polynomial.Factoring
  ( factor
  , rationalRoots
  , isIrreducible
  , factorSquareFree
  ) where

import Data.Ratio (numerator, denominator)
import Data.Maybe (mapMaybe)
import Math.Polynomial.Univariate
import Math.Internal.Positive (Positive)
import Math.Internal.PrimeFactors (factorise)

-- | Factor a polynomial over \(\mathbb{Q}\) into irreducible factors.
--
-- Returns a list of @(irreducible factor, multiplicity)@ pairs.
-- The factors are monic; any leading rational coefficient is separated out.
--
-- Uses 'squareFree' (Yun's algorithm) to separate multiplicities, then
-- 'factorSquareFree' on each square-free component.
--
-- >>> factor (mkPoly [1, 2, 1] :: Poly Rational)
-- [(Poly [1 % 1,1 % 1],2)]
factor :: Poly Rational -> [(Poly Rational, Int)]
factor p
  | degree p <= 0 = []
  | otherwise =
      let sfFactors = squareFree p
      in concatMap (\(f, m) -> map (\g -> (g, m)) (factorSquareFree f)) sfFactors

-- | Factor a square-free polynomial over \(\mathbb{Q}\) into irreducible factors.
--
-- Strategy:
--
-- 1. Extract linear factors via the rational root theorem.
-- 2. Attempt to find quadratic factors via Kronecker's method
--    (evaluation at three points and interpolation).
-- 3. If no factors are found, the polynomial is assumed irreducible.
--
-- Returns a list of monic irreducible factors.
factorSquareFree :: Poly Rational -> [Poly Rational]
factorSquareFree p
  | degree p <= 0 = []
  | degree p == 1 = [monicPoly p]
  | otherwise = go (monicPoly p)
  where
    go f
      | degree f <= 1 = [f]
      | otherwise =
          case findLinearFactor f of
            Just (root, quotient) ->
              mkPoly [-root, 1] : go quotient
            Nothing ->
              case findQuadraticFactor f of
                Just (fac, quotient) ->
                  fac : go quotient
                Nothing ->
                  -- Cannot factor further with current methods;
                  -- assume irreducible
                  [f]

-- | Find all rational roots of a polynomial using the rational root theorem.
--
-- If \(p(x) = a_n x^n + \cdots + a_0\) has integer coefficients (after
-- clearing denominators), then any rational root \(p/q\) (in lowest terms)
-- satisfies: \(p \mid a_0\) and \(q \mid a_n\).
--
-- >>> rationalRoots (mkPoly [-6, 11, -6, 1] :: Poly Rational)
-- [1 % 1,2 % 1,3 % 1]
rationalRoots :: Poly Rational -> [Rational]
rationalRoots p
  | degree p <= 0 = []
  | otherwise =
      let cs = unPoly p
          -- Clear denominators to get integer coefficients
          lcmDenom = foldl lcm 1 (map denominator cs)
          intCoeffs = map (\c -> numerator (c * fromInteger lcmDenom)) cs
          a0 = case intCoeffs of { (x:_) -> x; [] -> 0 }
          an = last intCoeffs
          -- Candidate numerators: divisors of a0
          -- Candidate denominators: divisors of an
          numDivs = if a0 == 0 then [0] else divisors (abs a0)
          denDivs = divisors (abs an)
          candidates = [fromInteger p' / fromInteger q'
                       | p' <- concatMap (\d -> [d, -d]) numDivs
                       , q' <- denDivs
                       , q' /= 0
                       ]
      in filter (\r -> evalPoly p r == 0) candidates

-- | Test whether a polynomial is irreducible over \(\mathbb{Q}\).
--
-- A polynomial is irreducible if it has degree \(\geq 1\) and cannot be
-- written as a product of two polynomials of lower degree over \(\mathbb{Q}\).
isIrreducible :: Poly Rational -> Bool
isIrreducible p
  | degree p <= 1 = degree p == 1
  | otherwise     = length (factorSquareFree p) == 1

-- Internal helpers

findLinearFactor :: Poly Rational -> Maybe (Rational, Poly Rational)
findLinearFactor p =
  case rationalRoots p of
    []    -> Nothing
    (r:_) -> Just (r, fst $ divModPoly p (mkPoly [-r, 1]))

-- | Try to find a quadratic factor using Kronecker's method.
--
-- Evaluates the polynomial at 0, 1, and -1 to obtain values, then
-- enumerates divisor triples and reconstructs candidate quadratic
-- factors via interpolation.
findQuadraticFactor :: Poly Rational -> Maybe (Poly Rational, Poly Rational)
findQuadraticFactor p
  | degree p < 4 = Nothing  -- degree 2 or 3 without linear factors is irreducible over Q...
                             -- well, degree 3 without linear factors is irreducible, degree 2 also
  | otherwise =
      -- Evaluate at 0, 1, -1 and try to find a quadratic factor
      -- whose values divide p's values at those points
      let v0 = evalPoly p 0
          v1 = evalPoly p 1
          vm1 = evalPoly p (-1)
          d0 = if v0 == 0 then [0] else rationalDivisors v0
          d1 = if v1 == 0 then [0] else rationalDivisors v1
          dm1 = if vm1 == 0 then [0] else rationalDivisors vm1
          -- For each triple (f(0), f(1), f(-1)), reconstruct the quadratic
          -- a + bx + cx^2 via interpolation:
          -- f(0) = a, f(1) = a+b+c, f(-1) = a-b+c
          -- => c = (f(1) + f(-1))/2 - a, b = (f(1) - f(-1))/2
          candidates = [ mkPoly [a, b, c]
                       | a <- d0
                       , f1 <- d1
                       , fm1 <- dm1
                       , let b = (f1 - fm1) / 2
                       , let c = (f1 + fm1) / 2 - a
                       , c /= 0
                       ]
      in case mapMaybe (tryDivide p) candidates of
           []          -> Nothing
           ((f,q):_)   -> Just (monicPoly f, q)

tryDivide :: Poly Rational -> Poly Rational -> Maybe (Poly Rational, Poly Rational)
tryDivide p f
  | degree f < 1 || degree f > degree p = Nothing
  | otherwise =
      let (q, r) = divModPoly p f
      in if unPoly r == [] then Just (f, q) else Nothing

-- | All positive divisors of a positive integer.
divisors :: Integer -> [Integer]
divisors 0 = [0]
divisors n =
  let fs = factorise (fromInteger n :: Positive)
  in go fs
  where
    go [] = [1]
    go ((p, e) : rest) =
      let ds = go rest
      in [p^k * d | k <- [0..e], d <- ds]

-- | Rational divisors: all rational numbers whose numerator divides the
-- given value's numerator and whose denominator divides its denominator,
-- with both signs.
rationalDivisors :: Rational -> [Rational]
rationalDivisors r =
  let n = abs (numerator r)
      d = abs (denominator r)
      nDivs = divisors n
      dDivs = divisors d
  in [fromInteger a / fromInteger b | a <- concatMap (\x -> [x, -x]) nDivs, b <- dDivs, b /= 0]
